test/validation/api/random/random.c

```diff
@@ -80,10 +80,370 @@ static void random_test_repeat(void)
+ * also odp_random_test_data(), set MIN_KIND to -1.
+ */
+#define MIN_KIND ODP_RANDOM_CRYPTO
+
+/*
+ * Alpha for P-value tests. This does not affect the tests that use a
+ * precomputed critical value.
+ */
+const double alpha = 0.00000001;
+
+static void random_data(uint8_t *buf, uint32_t len, odp_random_kind_t kind)
+{
+	static uint64_t seed;
+
+	switch (kind) {
+	case ODP_RANDOM_BASIC:
+	case ODP_RANDOM_CRYPTO:
+	case ODP_RANDOM_TRUE:
+		for (uint32_t i = 0; i < len;) {
+			int32_t r = odp_random_data(buf + i, len - i, kind);
+
+			CU_ASSERT_FATAL(r >= 0);
+			i += r;
+		}
+		break;
+	default:
+		CU_ASSERT_FATAL(odp_random_test_data(buf, len, &seed) ==
+				(int32_t)len);
+	}
+}
+
+static uint32_t random_bits(int n, odp_random_kind_t kind)
+{
+	static uint8_t buf[32 * 1024];
+	const int size = sizeof(buf);
+	static int cur_n;
+	static odp_random_kind_t cur_kind;
+	static int bit;
+	uint32_t r = 0;
+
+	if (n != cur_n || kind != cur_kind) {
+		cur_n = n;
+		cur_kind = kind;
+		bit = size * 8;
+	}
+
+	for (int i = 0; i < n; ) {
+		if (bit >= size * 8) {
+			random_data(buf, size, kind);
+			bit = 0;
+		}
+		if (n - i >= 8 && !(bit & 7)) {
+			/* Full byte. */
+			r <<= 8;
+			r |= buf[bit / 8];
+			bit += 8;
+			i += 8;
+			continue;
+		}
+		/* Single bit. */
+		r <<= 1;
+		r |= (buf[bit / 8] >> (7 - (bit & 7))) & 1;
+		bit++;
+		i++;
+	}
+
+	return r;
+}
+
+static const char *kname(odp_random_kind_t kind)
+{
+	static const char *kname_str[4] = {
+		"test",
+		"basic",
+		"crypto",
+		"true",
+	};
+
+	return kname_str[kind + 1];
+}
+
+/*
+ * Pearson's chi-squared goodness-of-fit test for uniform distribution. The test
+ * is run with multiple different bit block lengths. The null hypothesis is that
+ * each possible bit pattern is equally likely. If the chi-squared statistic is
+ * equal to or larger than the critical value, we conclude that the data is
+ * biased.
+ */
+static void random_test_frequency(void)
+{
+	/* Mean number of hits per cell. */
+	const uint32_t expected = 100;
+
+	/* From LibreOffice CHISQ.INV.RT(0.00000001; df). */
+	const double critical[] = {
+		32.8413, 40.1300, 50.8129, 68.0293,
+		97.0285, 147.463, 237.614, 402.685,
+		711.187, 1297.50, 2426.64, 4623.37,
+		8929.74, 17419.3, 34224.0, 67587.1,
+	};
+
+	printf("\n\n");
+
+	for (int kind = MIN_KIND; kind <= (int)odp_random_max_kind(); kind++) {
+		for (int bits = 1; bits <= 16; bits++) {
+			const uint32_t cells = 1 << bits;
+			const uint64_t num = expected * cells;
+			uint64_t f[256 * 256] = { 0 };
+
+			for (uint64_t i = 0; i < num; i++)
+				f[random_bits(bits, kind)]++;
+
+			double chisq = 0, crit = critical[bits - 1];
+
+			for (uint64_t i = 0; i < cells; i++) {
+				double dif = (double)f[i] - expected;
+
+				chisq += dif * dif / expected;
+			}
+
+			printf("%s ; bits %d ; chisq %g ; df %u ; crit %g ; ",
+			       kname(kind), bits, chisq, cells - 1, crit);
+
+			if (chisq < crit)
+				printf("pass\n");
+			else
+				printf("FAIL\n");
```

Comment:
A bit less code:
```c
printf("%s\n", chisq < crit ? "pass" : "FAIL");
```

Reply:
I put this in a function to boot, so that the strings appear in just one place.